### ISSUE TITLE
COMMAND_NOT_FOUND [get_gw_version()]

### ISSUE THREAD

**Issue by GD-Dal** [smar000/bisecur2mqtt#1]
Hi.
I know you mentioned that it will not be developed any further, but maybe someone else has encountered the same error as me:
When I run the script I get this error: COMMAND_NOT_FOUND [get_gw_version()]

The log says:

```
Connecting to MQTT broker [<module>()]
INIT: User 'hass' logged in to Bisecur Gateway with token '5429375' [do_gw_login()]
Garage door is OPEN [get_door_status()]
Connected to MQTT broker. Subscribing to 'bisecur2mqtt/send_command/command' [on_connect()]
Device responded with error! Code: 0 Reason: COMMAND_NOT_FOUND [get_gw_version()]
Exiting system. Changing MQTT state to 'offline' [<module>()]
```
So it seems to connect to the broker just fine. but that's it.

You also mention a `get_ports` command.. how do I run that?

Thanks

**Comment by smar000** [comment]
Hi

Not sure whey you are getting that error, as I have been running the script without issues for sometime now. Are you sure you have the [pysecur3](https://github.com/skelsec/pysecur3) library correctly installed?



> You also mention a `get_ports` command.. how do I run that?

Publish `get_ports` to your command topic.

**Comment by GD-Dal** [comment]
Hi, and thanks for answering.
Does pysecur3 require some additional installation?

I just cloned this github using
`git clone https://github.com/smar000/bisecur2mqtt.git`

And that gave me a file structure looking like this:
/home/supportpi/bisecur2mqtt/bisecur2mqtt

And inside that folder are your scripts + 2 pysecur3 folders:
/home/supportpi/bisecur2mqtt/bisecur2mqtt/pysecur3
/home/supportpi/bisecur2mqtt/bisecur2mqtt/pysecur3_tools

Does it need some extra installation beyond that? 

When I run the get_ports() function from your script, it gives me this:

```
Ports for user 0: [
    {
        "id": 0,
        "name": "Garasjeport",
        "ports": [
            {
                "id": 0,
                "type": 6
            },
            {
                "id": 1,
                "type": 4
            },
            {
                "id": 2,
                "type": 5
            },
            {
                "id": 3,
                "type": 8
            }
        ]
    }
```
Not really sure what it tells me, though.

**Comment by smar000** [comment]
> Does it need some extra installation beyond that?
> 
> When I run the get_ports() function from your script, it gives me this

If you are getting a response from `get_ports`, pysecur3 should be ok. From a quick look at the code, the command not found error you are seeing seems to be come from pysecur3 itself - bit unusual as it is working fine for me. Do you have any other versions of pysecur3 installed at the same time (maybe through a pip install or similar)?

Unfortunately I am a bit pushed for time at the moment and won't be able to help much further on this for at least the next month or two. For a temporary work around you could try commenting out the call to `get_gw_version()`, e.g. replace line 368 with:

`payload["gw_hw_version"] = "unknown"`

etc.

> Not really sure what it tells me, though.

The list of "ports" are basically telling what functions your bisecur door supports, e.g. up, down, stop etc. These need to be defined in your config file. E.g. for my door, I have the following in my config file:

```
impulse_port    = 0
up_port         = 3
down_port       = 4
partial_port    = 2
light_port      = 1
```

These were obtained from the following ports returned by `get_ports`:
```
            "ports": [
                {
                    "id": 1,
                    "type": 4
                },
                {
                    "id": 2,
                    "type": 5
                },
                {
                    "id": 3,
                    "type": 8
                },
                {
                    "id": 4,
                    "type": 6
                }
            ]
```
Hope that helps.

**Comment by GD-Dal** [comment]
Thanks for helping out.
The script called device_discovery.py works fine, and I got this info:
`Found device on address: 1.1.010.9, device attributes: {'swVersion': '2.5.0', 'hwVersion': '1.0.0', 'mac': '00:1e:f0:dd:a6:09', 'protocol': 'MCP V3.0'} `

So i changed line 368 from:
_, payload["gw_hw_version"] = get_gw_version()
to
payload["gw_hw_version"] = '1.0.0'

And and the script runs without errors now, which is fine.
All that remains now is to find out how to actually use it (in home assistant)

Thanks

**Comment by smar000** [comment]
Great to hear that you've got it running.

I don't use it in homeassistant myself (I am using openHAB), so can't really help you there. From my understanding of homeassistant, you would need to use the mqtt integration, and create `cover` entities.  I have tried to add homeassistant auto mqtt discovery into the code, but have not tested it in home assistant itself (again only in openHAB, which supports a limited version of the HA auto discovery format). If the code works correctly, your HA should automatically pick up the entity(ies).

**Comment by GD-Dal** [comment]
Thanks, but there are still things to do before it works properly.
It really bugs me when I use the "get_ports" command it gives me this:

```
ports": [
            {
                "id": 0,
                "type": 6
            },
            {
                "id": 1,
                "type": 4
            },
            {
                "id": 2,
                "type": 5
            },
            {
                "id": 3,
                "type": 8
            }
        ]
    }
```
There is a id: 0 in there, which is the same port as IMPULSE.
The "partial" command does not seem to work (maybe because it has id: 0?)
The "up" and "down" command seems to do the same thing. If I publish "up" while the port is down, it opens. If I publish "up" again while the port is open, it closes. Same with the "down" command.
The only command I know for sure works as intended, is the "light" command (id: 3). But is there a way to see the state of the light?

The home assistant integration does not work properly. The device is recognized as a single entity, and not as a device with multiple entities. 
I will try to look into this.

Thanks again.

**Comment by smar000** [comment]
Sorry, I have no idea why your commands seem to be all over the place. It may well be that your gateway is on a later firmware than the one used by the pysecur3. Other than that, I can't really think of anything as it is working on my system. 

Unfortunately I don't think I can help you further on this at this time. I do hope though that you manage to get it working. Good luck!

**Comment by GD-Dal** [comment]
Yeah, I'm close to giving up on this one. I have been staring at this script for a week without any solution.
Not being able to make the MQTT Discovery thing working is one thing, but the main script crashes and loses contact with the gateway. 
Especially when the stop button is involved it will crash.

Are you sure the IMPULSE is not possible to move away from 0? But doing it directly in the pysecur3 library somehow?

**Comment by smar000** [comment]
I would go back to basics and make sure that the pysecur3 library can actually control your unit correctly. Once that is working, the mqtt wrapper bit is very straight forward. Crashes can be easily caught - put a try/except block around wherever the crash occurs to catch the error and handle it gracefully. Is the crash in pysecur3 or in my wrapper? etc. etc.

> Are you sure the IMPULSE is not possible to move away from 0? But doing it directly in the pysecur3 library somehow?

I have no idea. I am not doing anything other putting an mqtt wrapper around pysecur. All the controlling etc is being done there... If I had to guess, I would say that the the IDs for impulse/up/down etc are fixed by Bisecur themselves.
